sdk/src/main/java/com/karhoo/sdk/analytics/Payloader.kt

```diff
@@ -144,19 +176,140 @@ class Payloader internal constructor(builder: Builder) {
-                          date: Date,
-                          amount: Int,
-                          currency: String): Builder {
+        fun cardAuthorisationFailure(
+            errorMessage: String,
+            quoteId: String?,
+            paymentMethodUsed: String,
+            lastFourDigits: String,
+            date: Date,
+            amount: Int,
+            currency: String
+        ): Builder {
+            quoteId?.let {
+                payload[QUOTE_ID] = it
+            }
+
             payload[PAYMENT_ERROR_MESSAGE] = errorMessage
             payload[PAYMENT_CARD_LAST_FOUR_DIGITS] = lastFourDigits
             payload[PAYMENT_AMOUNT] = amount
             payload[PAYMENT_CURRENCY] = currency
+            payload[PAYMENT_METHOD_USED] = paymentMethodUsed
             payload[PAYMENT_DATE] = date.toString()
             return this
         }
 
+        fun checkoutOpened(quoteId: String?): Builder {
+            quoteId?.let {
+                payload[QUOTE_ID] = it
+            }
+
+            return this
+        }
+
+        fun cardAuthorisationSuccess(quoteId: String?): Builder {
+            quoteId?.let {
+                payload[QUOTE_ID] = it
+            }
+
+            return this
+        }
+
+        fun loyaltyStatusRequested(
+            slug: String?,
+            errorMessage: String?,
+            quoteId: String?,
+            correlationId: String?,
+            loyaltyStatus: LoyaltyStatus?,
+            balance: Int?,
+            loyaltyProgramme: LoyaltyProgramme?
+        ): Builder {
+            payload[LOYALTY_ENABLED] =
+                loyaltyStatus?.canBurn == true || loyaltyStatus?.canEarn == true
+
+            payload[LOYALTY_STATUS_SUCCESS] = loyaltyStatus != null
+            loyaltyStatus?.let {
+                it.canBurn?.let { canBurn ->
+                    payload[LOYALTY_STATUS_CAN_BURN] = canBurn
+                }
+                it.canEarn?.let { canEarn ->
+                    payload[LOYALTY_STATUS_CAN_EARN] = canEarn
+                }
+                balance?.let { balance ->
+                    payload[LOYALTY_STATUS_BALANCE] = balance
+                }
+            }
+
+            quoteId?.let {
+                payload[QUOTE_ID] = it
+            }
+            loyaltyProgramme?.name?.let {
+                payload[LOYALTY_NAME] = it
+            }
+            correlationId?.let {
+                payload[CORRELATION_ID] = it
+            }
+            slug?.let {
+                payload[LOYALTY_STATUS_ERROR_SLUG] = it
+            }
+            errorMessage?.let {
+                payload[LOYALTY_STATUS_ERROR_MESSAGE] = errorMessage
+            }
+
+            return this
+        }
+
+        fun loyaltyPreAuthFailure(
+            slug: String?,
+            errorMessage: String?,
+            quoteId: String?,
+            correlationId: String?,
+            loyaltyMode: String,
+            balance: Int?
```

Comment:
is the `balance` param present on iOS as well @BartlomiejSopalaKarhoo? it's not in the excel so not sure if the excel is outdated here or if this is an oversight.

Reply:
@mionescu95 same as above, why are most of the params nullable? is there a situation where we need to fire this event, but don't have the `quoteId` for example?

Reply:
Removed the balance, kept the loyaltyMode